File: src/types/Lessons_Guideline.ts
/**
 * @main Guideline
 */

import { TypeConfig } from "../typeConfig.js"
import { todo } from "../validation/builders/integrity.js"
import { validateEntityFileName } from "../validation/builders/naming.js"
import { createSchemaValidator } from "../validation/builders/schema.js"
import { getFilenamePrefixAsNumericId } from "../validation/filename.js"
import { LocaleMap } from "./_LocaleMap.js"
import { NonEmptyString } from "./_NonEmptyString.js"

/**
 * @title Guideline
 */
export type Guideline = {
  /**
   * The guideline's identifier. An unique, increasing integer.
   * @integer
   * @minimum 1
   */
  id: number

  /**
   * Maximum number of spells that can be exchanged.
   * @integer
   * @minimum 0
   */
  spellwork_changes_allowed: number

  /**
   * All translations for the entry, identified by IETF language tag (BCP47).
   */
  translations: LocaleMap<GuidelineTranslation>
}

export type GuidelineTranslation = {
  /**
   * The guideline name.
   */
  name: NonEmptyString
}

export const config: TypeConfig<Guideline> = {
  name: "Guideline",
  id: getFilenamePrefixAsNumericId,
  integrityValidator: todo("Guideline"),
  schemaValidator: createSchemaValidator(import.meta.url),
  fileNameValidator: validateEntityFileName,
}

File: Sources/OptolithDatabaseSchema/GeneratedTypes/Lessons_Guideline.swift
//
//  Lessons_Guideline.swift
//  OptolithDatabaseSchema
//

import DiscriminatedEnum

public struct Guideline: LocalizableEntity {
    /// The guideline's identifier. An unique, increasing integer.
    public let id: Int

    /// Maximum number of spells that can be exchanged.
    public let spellworkChangesAllowed: Int

    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<GuidelineTranslation>

    public init(id: Int, spellworkChangesAllowed: Int, translations: LocaleMap<GuidelineTranslation>) {
        self.id = id
        self.spellworkChangesAllowed = spellworkChangesAllowed
        self.translations = translations
    }

    private enum CodingKeys: String, CodingKey {
        case id = "id"
        case spellworkChangesAllowed = "spellwork_changes_allowed"
        case translations = "translations"
    }
}

public struct GuidelineTranslation: EntitySubtype {
    /// The guideline name.
    public let name: NonEmptyString

    public init(name: NonEmptyString) {
        self.name = name
    }
}
